A2-service-configs-in-dns.md

```diff
@@ -0,0 +1,129 @@
+Service Config via DNS
+----------------------
+* Author(s): Mark D. Roth (roth@google.com)
+* Approver: a11r
+* Status: Draft
+* Implemented in: N/A
+* Last updated: 2017-01-19
+* Discussion at: https://groups.google.com/d/topic/grpc-io/DkweyrWEXxU/discussion
+
+## Abstract
+
+This document proposes a mechanism for encoding gRPC service config data
+in DNS for use in the open-source world.
+
+## Background
+
+The [service
+config](https://github.com/grpc/grpc/blob/master/doc/service_config.md)
+mechanism was originally designed for use inside of Google.  However,
+all but one part of the original design will work fine in the open-source
+world.  That one part is the specification of how the service config
+data will be encoded in DNS.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+There are two parts to this proposal.  The first part is to add some
+JSON wrapping for controlling how service config changes are canary
+tested.  The second part describes how the service config is encoded in
+DNS.
+
+### Canarying Changes
+
+When deploying a change to a service config, it is useful to be able to
+canary test changes to avoid wide-spread breakage by slowly increasing the
+number of clients that see the new version.  To that end, multiple
+service configs choices can be listed, in order, along with criteria that
+determine which choice will be selected by a given client:
+
+```
+// A list of one or more service config choices.
+// The first matching entry wins.
+[
+  {
+    // Criteria used to select this choice.
+    // If a field is absent or empty, it matches all clients.
+    // All fields must match a client for this choice to be selected.
+    //
+    // Client language(s): a list of strings (e.g., 'c++', 'java', 'go',
+    // 'python', etc).
+    'clientLanguage': [string],
+    // Percentage: integer from 0 to 100 indicating the percentage of
+    // clients that should use this choice.
+    'percentage': number,
+    // Client hostname(s): a list of strings.
```

Comment:
The client hostname will be an exact match.  In the discussion thread, we considered making this be a wildcard match of some sort, but it's not clear that there's actually a use-case for that right now, so we decided to just stick with exact-match semantics for now.  We can always add a new type of selection criteria later, if/when we have a real use-case for it.

I'm not sure what you mean by a client getting an explicit name other than a hostname.  We could certainly add something like that later, but I'd like to see a real use-case for it before doing so.